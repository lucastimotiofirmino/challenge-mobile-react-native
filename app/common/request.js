import axios from 'axios';
import md5 from 'crypto-js/md5';

import { MARVEL_API_URL, MARVEL_API_PUBLIC, MARVEL_API_PRIVATE } from '@env';

export const axiosInstance = axios.create({
  baseURL: MARVEL_API_URL,
});

const JSON_HEADER = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
};

export function makeRequestParams({ method, body }) {
  return {
    method,
    headers: {
      ...JSON_HEADER,
    },
    ...(body !== null && { data: JSON.stringify(body) }),
  };
}

export function makeApiHash({ timestamp, privateKey, publicKey }) {
  return md5(`${timestamp}${privateKey}${publicKey}`);
}

export function appendHashToUrl({ hash, publicKey, timestamp }) {
  return `&ts=${timestamp}&apikey=${publicKey}&hash=${hash}`;
}

async function request({
  url,
  method = 'GET',
  body = null,
  timestamp = new Date().getTime(),
}) {
  const requestParams = makeRequestParams({
    method,
    body,
  });

  const hash = makeApiHash({
    timestamp,
    publicKey: MARVEL_API_PUBLIC,
    privateKey: MARVEL_API_PRIVATE,
  });

  const appendUrl = appendHashToUrl({
    timestamp,
    hash,
    publicKey: MARVEL_API_PUBLIC,
  });

  return axiosInstance
    .request({ url: `${url}${appendUrl}`, ...requestParams })
    .then((response) => response);
}

export default request;
