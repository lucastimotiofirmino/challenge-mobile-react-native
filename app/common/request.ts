import axios, { Method } from 'axios';
import md5 from 'crypto-js/md5';

import { MARVEL_API_URL, MARVEL_API_PUBLIC, MARVEL_API_PRIVATE } from '@env';

export const axiosInstance = axios.create({
  baseURL: MARVEL_API_URL,
});

const JSON_HEADER = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
};

export function makeRequestParams(method: Method) {
  return {
    method,
    headers: {
      ...JSON_HEADER,
    },
  };
}

export function makeApiHash(
  timestamp: number,
  privateKey: string,
  publicKey: string,
) {
  return md5(`${timestamp}${privateKey}${publicKey}`).toString();
}

export function appendHashToUrl(
  timestamp: number,
  publicKey: string,
  hash: string,
) {
  return `&ts=${timestamp}&apikey=${publicKey}&hash=${hash}`;
}

interface IRequest {
  url: string;
  method?: Method;
  timestamp?: number;
}

async function request({
  url,
  method = 'GET',
  timestamp = new Date().getTime(),
}: IRequest) {
  const requestParams = makeRequestParams(method);
  const hash = makeApiHash(timestamp, MARVEL_API_PRIVATE, MARVEL_API_PUBLIC);
  const appendUrl = appendHashToUrl(timestamp, MARVEL_API_PUBLIC, hash);

  return axiosInstance
    .request({ url: `${url}${appendUrl}`, ...requestParams })
    .then((response) => response);
}

export default request;
