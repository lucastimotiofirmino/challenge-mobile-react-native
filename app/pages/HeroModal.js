import React, { memo } from 'react';

import { View, StyleSheet, Text, Image } from 'react-native';

import BaseModal from '../components/BaseModal';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'center',
  },
  side: {
    maxWidth: 100,
  },
  avatar: {
    width: 100,
    height: 100,
    borderRadius: 50,
  },
  name: {
    fontWeight: 'bold',
    textAlign: 'center',
    marginVertical: 5,
    fontSize: 18,
  },
  descriptionBox: {
    width: 220,
    marginLeft: 20,
  },
  divider: {
    height: StyleSheet.hairlineWidth,
    backgroundColor: '#999',
    marginVertical: 10,
  },
});

const HeroModal = ({ route }) => {
  const { name, description, avatar, numEvents } = route.params;
  return (
    <BaseModal>
      <View style={styles.container}>
        <View style={styles.side}>
          <Image source={{ uri: avatar }} style={styles.avatar} />
          <Text style={styles.name}>{name}</Text>
        </View>
        <View style={styles.descriptionBox}>
          <Text>{description || 'Nenhuma descrição disponível.'}</Text>

          {numEvents > 0 && (
            <>
              <View style={styles.divider} />
              {numEvents > 0 && (
                <Text>
                  Aparece em
                  {numEvents == 1 ? ' um evento' : ` ${numEvents} eventos`}.
                </Text>
              )}
            </>
          )}
        </View>
      </View>
    </BaseModal>
  );
};

export default memo(HeroModal);
