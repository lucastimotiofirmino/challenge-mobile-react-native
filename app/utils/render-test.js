import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import createStore from '../common/store';

export function renderWithRedux(ui, { store, ...otherOpts }) {
  return render(<Provider store={store}>{ui}</Provider>, otherOpts);
}

export function makeTestStore(opts = {}) {
  const { store } = createStore(opts);
  return store;
}
