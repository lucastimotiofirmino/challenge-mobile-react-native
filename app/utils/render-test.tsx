import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import createStore from '../common/store';
import { Store } from 'redux';

export function renderWithRedux(
  ui: React.ReactNode,
  { store, ...otherOpts }: { store: Store },
) {
  return render(<Provider store={store}>{ui}</Provider>, otherOpts);
}

export function makeTestStore() {
  const { store } = createStore();
  return store;
}
