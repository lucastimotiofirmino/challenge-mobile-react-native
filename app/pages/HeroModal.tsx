import React, { memo } from 'react';

import {
  View,
  StyleSheet,
  Text,
  Image,
  ViewStyle,
  TextStyle,
  ImageStyle,
} from 'react-native';
import { IHero } from '../common/interfaces';

import BaseModal from '../components/BaseModal';

type Style = {
  container: ViewStyle;
  side: ViewStyle;
  avatar: ImageStyle;
  name: TextStyle;
  descriptionBox: ViewStyle;
  divider: ViewStyle;
};

const styles = StyleSheet.create<Style>({
  container: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'center',
  },
  side: {
    maxWidth: 100,
  },
  avatar: {
    width: 100,
    height: 100,
    borderRadius: 50,
  },
  name: {
    fontWeight: 'bold',
    textAlign: 'center',
    marginVertical: 5,
    fontSize: 18,
  },
  descriptionBox: {
    width: 220,
    marginLeft: 20,
  },
  divider: {
    height: StyleSheet.hairlineWidth,
    backgroundColor: '#999',
    marginVertical: 10,
  },
});

type HeroModalProps = {
  route: {
    params: IHero;
  };
};

const HeroModal: React.FC<HeroModalProps> = ({ route }) => {
  const { name, description, avatar, numEvents } = route.params;
  return (
    <BaseModal>
      <View style={styles.container}>
        <View style={styles.side}>
          <Image source={{ uri: avatar }} style={styles.avatar} />
          <Text style={styles.name}>{name}</Text>
        </View>
        <View style={styles.descriptionBox}>
          <Text>{description || 'Nenhuma descrição disponível.'}</Text>

          {numEvents > 0 && (
            <>
              <View style={styles.divider} />
              {numEvents > 0 && (
                <Text>
                  Aparece em
                  {numEvents === 1 ? ' um evento' : ` ${numEvents} eventos`}.
                </Text>
              )}
            </>
          )}
        </View>
      </View>
    </BaseModal>
  );
};

export default memo(HeroModal);
