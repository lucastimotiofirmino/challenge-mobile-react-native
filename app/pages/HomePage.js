import React, { useState, useEffect, useRef } from 'react';
import {
  StyleSheet,
  View,
  RefreshControl,
  FlatList,
  Text,
  Alert,
} from 'react-native';
import { SearchBar } from 'react-native-elements';
import { useDispatch, useSelector } from 'react-redux';
import Icon from 'react-native-vector-icons/AntDesign';

import Loader from '../components/Loader';
import Field from '../components/Field';
import HeroCard, { CARD_HEIGHT } from '../components/HeroCard';

import { getHeroes } from '../actions/heroes';

import { filterWithQuerySearch } from '../common/helpers';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  info: {
    textAlign: 'center',
    marginBottom: 10,
  },
  flatList: {
    alignItems: 'center',
    paddingBottom: 200,
  },
  filters: {
    flexDirection: 'row',
    marginHorizontal: 10,
    marginVertical: 10,
  },
  endText: {
    marginTop: 10,
    marginBottom: 200,
    fontSize: 24,
  },
});

const CARD_LIMIT_PER_BATCH = 10;

const renderHeroCard = ({ item }) => <HeroCard {...item} />;
const heroKeyExtractor = (item) => `hero-${item.id}`;

const searchIcon = <Icon name="filter" size={24} color="#666" />;
const endList = <Text style={styles.endText}>That's all folks!!!</Text>;
const loader = <Loader />;

const InfoTextBar = ({ content }) => <Text style={styles.info}>{content}</Text>;

const HomePage = () => {
  const dispatch = useDispatch();
  const flatListRef = useRef();

  const [searchText, setSearchText] = useState('');

  const [loading, setLoading] = useState(false);
  const [filtersParams, setFilterParam] = useState({
    offset: 0,
    alphabetOrder: true,
    searchQuery: '',
  });

  const [onlyFavorites, setFavorites] = useState(false);

  const heroesNormal = useSelector((state) => state.heroes.heroes);
  const heroesNormalTotal = useSelector((state) => state.heroes.total);
  const hasNextPage = useSelector((state) => state.heroes.hasNext);

  // this below is stored as local
  const hereosFavorites = useSelector((state) => state.heroes.favorites);

  const heroes = onlyFavorites
    ? filterWithQuerySearch(hereosFavorites, searchText)
    : heroesNormal;
  const total = onlyFavorites ? heroes.length : heroesNormalTotal;

  resetList = () => {
    setLoading(false);
    setSearchText('');
    setFilterParam({ ...filtersParams, offset: 0, searchQuery: '' });
  };

  changeAndFetchOrder = async () => {
    setFilterParam({
      ...filtersParams,
      offset: 0,
      alphabetOrder: !filtersParams.alphabetOrder,
    });
  };

  changeToFavoritesOrNot = () => {
    setFavorites(!onlyFavorites);
    if (flatListRef.current) {
      flatListRef.current.scrollToOffset({ animated: false, offset: 0 });
    }
  };

  fetchSearch = () => {
    setFilterParam({ ...filtersParams, offset: 0, searchQuery: searchText });
  };

  fetchMore = () => {
    if (!loading && hasNextPage && searchText === '' && !onlyFavorites) {
      setFilterParam({
        ...filtersParams,
        offset: filtersParams.offset + CARD_LIMIT_PER_BATCH,
      });
    }
  };

  useEffect(() => {
    if (onlyFavorites) return;

    setLoading(true);

    dispatch(
      getHeroes({
        offset: filtersParams.offset,
        name: filtersParams.searchQuery,
        orderBy: filtersParams.alphabetOrder ? 'name' : '-name',
      }),
    )
      .catch((error) => Alert.alert(error.message))
      .done(() => {
        setLoading(false);
      });
  }, [filtersParams]);

  return (
    <View styles={styles.container}>
      <SearchBar
        platform={Platform.OS}
        onChangeText={(value) => setSearchText(value)}
        searchIcon={searchIcon}
        cancelIcon={searchIcon}
        onSubmitEditing={() => fetchSearch()}
        returnKeyType="search"
        clearIcon={
          <Icon
            name="close"
            size={18}
            color="#666"
            onPress={() => resetList()}
          />
        }
        cancelButtonTitle="Cancelar"
        placeholder="Filtre um herói da Marvel pelo nome"
        value={searchText}
      />

      <View style={styles.filters}>
        <Field
          text="Meus favoritos"
          icon="star"
          onPress={() => changeToFavoritesOrNot()}
          backgroundColor={onlyFavorites ? '#2E8C79' : '#44CBB1'}
        />
        {!onlyFavorites && (
          <Field
            text={
              filtersParams.alphabetOrder
                ? 'Ordem alfabética (a-z)'
                : 'Ordem alfabética (z-a)'
            }
            backgroundColor="#44CBB1"
            icon={filtersParams.alphabetOrder ? 'arrowdown' : 'arrowup'}
            onPress={() => changeAndFetchOrder()}
          />
        )}
      </View>

      <FlatList
        ref={flatListRef}
        keyboardDismissMode="on-drag"
        showsVerticalScrollIndicator={false}
        numColumns={2}
        data={heroes}
        contentContainerStyle={styles.flatList}
        onEndReached={fetchMore}
        onEndReachedThreshold={0.9}
        extraData={[heroes, loading, searchText]}
        renderItem={renderHeroCard}
        keyExtractor={heroKeyExtractor}
        refreshControl={
          <RefreshControl
            refreshing={false}
            color="#000"
            onRefresh={() => resetList()}
          />
        }
        ListHeaderComponent={
          total > 0 && <InfoTextBar content={`${total} heróis`} />
        }
        ListEmptyComponent={
          total <= 0 &&
          !loading && <InfoTextBar content="Nenhum herói encontrado." />
        }
        ListFooterComponent={
          loading && hasNextPage
            ? loader
            : !hasNextPage && filtersParams.searchQuery === '' && endList
        }
        getItemLayout={(data, index) => ({
          length: CARD_HEIGHT,
          offset: CARD_HEIGHT * index,
          index,
        })}
      />
    </View>
  );
};

export default HomePage;
