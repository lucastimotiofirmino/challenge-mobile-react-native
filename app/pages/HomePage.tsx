import React, { useState, useEffect, useRef } from 'react';
import {
  StyleSheet,
  View,
  RefreshControl,
  FlatList,
  Text,
  Alert,
  Platform,
  ViewStyle,
  FlexStyle,
  TextStyle,
} from 'react-native';
import { SearchBar } from 'react-native-elements';
import { useDispatch, useSelector } from 'react-redux';
import Icon from 'react-native-vector-icons/AntDesign';

import Loader from '../components/Loader';
import Field from '../components/Field';
import HeroCard, { CARD_HEIGHT } from '../components/HeroCard';

import { getHeroes } from '../actions/heroes';

import { filterWithQuerySearch } from '../common/helpers';
import { RootState } from '../reducers';
import { IHero } from '../common/interfaces';

type Style = {
  info: TextStyle;
  flatList: ViewStyle | FlexStyle;
  filters: ViewStyle | FlexStyle;
  endText: TextStyle;
};

const styles = StyleSheet.create<Style>({
  info: {
    textAlign: 'center',
    marginBottom: 10,
  },
  flatList: {
    alignItems: 'center',
    paddingBottom: 200,
  },
  filters: {
    flexDirection: 'row',
    marginHorizontal: 10,
    marginVertical: 10,
  },
  endText: {
    marginTop: 10,
    marginBottom: 200,
    fontSize: 24,
  },
});

const CARD_LIMIT_PER_BATCH: number = 10;

const renderHeroCard = ({ item }: { item: IHero }) => <HeroCard {...item} />;
const heroKeyExtractor = (item: IHero) => `hero-${item.id}`;

const searchIcon = <Icon name="filter" size={24} color="#666" />;
const endList = <Text style={styles.endText}>That's all folks!!!</Text>;
const loader = <Loader />;

const InfoTextBar = ({ content }: { content: string }) => (
  <Text style={styles.info}>{content}</Text>
);

const HomePage: React.FC = () => {
  const dispatch = useDispatch();
  const flatListRef = useRef();

  const [searchText, setSearchText] = useState<string>('');

  const [loading, setLoading] = useState<boolean>(false);
  const [filtersParams, setFilterParam] = useState({
    offset: 0,
    alphabetOrder: true,
    searchQuery: '',
  });

  const [onlyFavorites, setFavorites] = useState<boolean>(false);

  const heroesNormal = useSelector((state: RootState) => state.heroes.heroes);
  const heroesNormalTotal = useSelector(
    (state: RootState) => state.heroes.total,
  );
  const hasNextPage = useSelector((state: RootState) => state.heroes.hasNext);

  // this below is stored as local
  const hereosFavorites = useSelector(
    (state: RootState) => state.heroes.favorites,
  );

  const heroes = onlyFavorites
    ? filterWithQuerySearch(hereosFavorites, searchText)
    : heroesNormal;
  const total = onlyFavorites ? heroes.length : heroesNormalTotal;

  const resetList = () => {
    setLoading(false);
    setSearchText('');
    setFilterParam({ ...filtersParams, offset: 0, searchQuery: '' });
  };

  const changeAndFetchOrder = async () => {
    setFilterParam({
      ...filtersParams,
      offset: 0,
      alphabetOrder: !filtersParams.alphabetOrder,
    });
  };

  const changeToFavoritesOrNot = () => {
    setFavorites(!onlyFavorites);
    if (flatListRef.current) {
      flatListRef.current.scrollToOffset({ animated: false, offset: 0 });
    }
  };

  const fetchSearch = () => {
    setFilterParam({ ...filtersParams, offset: 0, searchQuery: searchText });
  };

  const fetchMore = () => {
    if (!loading && hasNextPage && searchText === '' && !onlyFavorites) {
      setFilterParam({
        ...filtersParams,
        offset: filtersParams.offset + CARD_LIMIT_PER_BATCH,
      });
    }
  };

  useEffect(() => {
    if (onlyFavorites) return;

    setLoading(true);

    dispatch(
      getHeroes({
        offset: filtersParams.offset,
        name: filtersParams.searchQuery,
        orderBy: filtersParams.alphabetOrder ? 'name' : '-name',
      }),
    )
      .catch((error: {message: string}) => Alert.alert(error.message))
      .done(() => {
        setLoading(false);
      });
  }, [filtersParams]);

  return (
    <>
      <SearchBar
        platform={Platform.OS}
        onChangeText={(value) => setSearchText(value)}
        searchIcon={searchIcon}
        cancelIcon={searchIcon}
        onSubmitEditing={() => fetchSearch()}
        returnKeyType="search"
        clearIcon={
          <Icon
            name="close"
            size={18}
            color="#666"
            onPress={() => resetList()}
          />
        }
        cancelButtonTitle="Cancelar"
        placeholder="Filtre um herói da Marvel pelo nome"
        value={searchText}
      />

      <View style={styles.filters}>
        <Field
          text="Meus favoritos"
          icon="star"
          onPress={() => changeToFavoritesOrNot()}
          backgroundColor={onlyFavorites ? '#2E8C79' : '#44CBB1'}
        />
        {!onlyFavorites && (
          <Field
            backgroundColor="#44CBB1"
            icon={filtersParams.alphabetOrder ? 'arrowdown' : 'arrowup'}
            onPress={() => changeAndFetchOrder()}
            text={
              filtersParams.alphabetOrder
                ? 'Ordem alfabética (a-z)'
                : 'Ordem alfabética (z-a)'
            }
          />
        )}
      </View>

      <FlatList
        ref={() => flatListRef}
        keyboardDismissMode="on-drag"
        showsVerticalScrollIndicator={false}
        numColumns={2}
        data={heroes}
        contentContainerStyle={styles.flatList}
        onEndReached={fetchMore}
        onEndReachedThreshold={0.9}
        extraData={[heroes, loading, searchText]}
        renderItem={renderHeroCard}
        keyExtractor={heroKeyExtractor}
        refreshControl={
          <RefreshControl refreshing={false} onRefresh={() => resetList()} />
        }
        ListHeaderComponent={
          total > 0 ? <InfoTextBar content={`${total} heróis`} /> : null
        }
        ListEmptyComponent={
          total <= 0 && !loading ? (
            <InfoTextBar content="Nenhum herói encontrado." />
          ) : null
        }
        ListFooterComponent={
          loading && hasNextPage
            ? loader
            : !hasNextPage && filtersParams.searchQuery === ''
            ? endList
            : null
        }
        getItemLayout={(data, index) => ({
          length: CARD_HEIGHT,
          offset: CARD_HEIGHT * index,
          index,
        })}
      />
    </>
  );
};

export default HomePage;
