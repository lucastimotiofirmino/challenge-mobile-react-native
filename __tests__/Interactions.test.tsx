/* eslint-disable no-undef */
import React from 'react';
import { fireEvent, waitFor } from '@testing-library/react-native';
import { useNavigation } from '@react-navigation/native';
import { renderWithRedux, makeTestStore } from '../app/utils/render-test';
import HomePage from '../app/pages/HomePage';
import HeroModal from '../app/pages/HeroModal';

import { axiosInstance } from '../app/common/request';
import MockAdapter from 'axios-mock-adapter';
const mockApi = new MockAdapter(axiosInstance);

const Wolverine = {
  id: "1",
  name: 'Wolverine',
  avatar: 'test.jpg',
  thumbnail: {
    path: 'test',
    extension: 'jpg',
  },
  description: 'Test description',
  events: { available: 1 },
  numEvents: 1,
};

const Abyss = {
  id: "2",
  name: 'Abyss',
  avatar: 'test.jpg',
  thumbnail: {
    path: 'test',
    extension: 'jpg',
  },
  events: { available: 1 },
  numEvents: 1,
};

describe('HomePage interaction suite', () => {
  const realStore = makeTestStore();
  const mockNavigate = jest.fn();
  const mocksetOptions = jest.fn();

  beforeEach(() => {
    //@ts-ignore
    useNavigation.mockImplementation(() => ({
      navigate: mockNavigate,
      setOptions: mocksetOptions,
    }));
  });

  test('renders the heroes homepage initial', async () => {
    mockApi.onGet(/characters*/).replyOnce(200, {
      data: {
        total: 0,
        count: 0,
        results: [Wolverine, Abyss],
      },
    });
    const { getByText } = renderWithRedux(<HomePage />, {
      store: realStore,
    });

    await waitFor(() => expect(getByText(/Wolverine/i)).toBeTruthy());
    await waitFor(() => expect(getByText(/Abyss/i)).toBeTruthy());
  });

  test('user can search heroes', async () => {
    mockApi.onGet(/characters*/).replyOnce(200, {
      data: {
        total: 0,
        count: 0,
        results: [Wolverine],
      },
    });

    const { getByText, getByPlaceholderText } = renderWithRedux(<HomePage />, {
      store: realStore,
    });

    mockApi.onGet(/nameStartsWith*/).replyOnce(200, {
      data: {
        total: 0,
        count: 0,
        results: [Abyss],
      },
    });

    const input = getByPlaceholderText(/Filtre um herÃ³i/i);
    await fireEvent.changeText(input, 'aby');
    await fireEvent(input, 'onSubmitEditing');

    await waitFor(() => expect(getByText(/Abyss/i)).toBeTruthy());
  });

  test('user can change the alfabetic order', async () => {
    mockApi.onGet(/characters*/).replyOnce(200, {
      data: {
        total: 0,
        count: 0,
        results: [Wolverine],
      },
    });

    const { queryAllByTestId, getByText } = renderWithRedux(<HomePage />, {
      store: realStore,
    });

    const cards = queryAllByTestId(/fieldbutton/);
    const secondFieldAlpabetic = cards[1];

    mockApi.onGet(/orderBy\=\-name*/).replyOnce(200, {
      data: {
        total: 0,
        count: 0,
        results: [Abyss],
      },
    });

    await fireEvent.press(secondFieldAlpabetic);

    await waitFor(() => expect(getByText(/Abyss/i)).toBeTruthy());
  });

  test('user can open hero modal', async () => {
    mockApi.onGet(/characters*/).replyOnce(200, {
      data: {
        total: 0,
        count: 0,
        results: [Wolverine],
      },
    });

    const { queryAllByTestId, getByText } = renderWithRedux(<HomePage />, {
      store: realStore,
    });

    await waitFor(() => expect(getByText(/Wolverine/i)).toBeTruthy());

    const cards = queryAllByTestId(/cardHero/);

    const firstCard = cards[0];

    await fireEvent.press(firstCard);

    await waitFor(() => expect(mockNavigate).toHaveBeenCalledTimes(1));
    expect(mockNavigate).toHaveBeenCalledWith('HeroModal', {
      ...Wolverine,
      avatar: 'test.jpg',
      favorite: false,
      numEvents: 1,
    });
  });

  test('user see hero modal', async () => {
    const { getByText, toJSON } = renderWithRedux(
      <HeroModal route={{ params: Wolverine }} />,
      { store: realStore },
    );

    await waitFor(() => expect(getByText(/Wolverine/i)).toBeTruthy());
    await waitFor(() => expect(getByText(/Test description/i)).toBeTruthy());

    expect(toJSON()).toMatchSnapshot();
  });
});
