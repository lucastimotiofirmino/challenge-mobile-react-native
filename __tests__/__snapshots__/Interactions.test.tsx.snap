// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage interaction suite user see hero modal 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Object {
          "alignSelf": "flex-end",
          "color": "#fff",
          "fontSize": 32,
          "marginBottom": 10,
          "marginHorizontal": 15,
        },
        Object {
          "fontFamily": "anticon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    î˜³
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 20,
        "justifyContent": "center",
        "paddingHorizontal": 10,
        "paddingVertical": 50,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "wrap",
        }
      }
    >
      <View
        style={
          Object {
            "maxWidth": 100,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "test.jpg",
            }
          }
          style={
            Object {
              "borderRadius": 50,
              "height": 100,
              "width": 100,
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginVertical": 5,
              "textAlign": "center",
            }
          }
        >
          Wolverine
        </Text>
      </View>
      <View
        style={
          Object {
            "marginLeft": 20,
            "width": 220,
          }
        }
      >
        <Text>
          Test description
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#999",
              "height": 0.5,
              "marginVertical": 10,
            }
          }
        />
        <Text>
          Aparece em
           um evento
          .
        </Text>
      </View>
    </View>
  </View>
</View>
`;
